TASVideos/Pages/Search/Advanced.cshtml.cs

```diff
@@ -0,0 +1,163 @@
+
+	public List<PageSearchModel> PageResults { get; set; } = new();
+	public List<TopicSearchModel> TopicResults { get; set; } = new();
+	public List<PostSearchModel> PostResults { get; set; } = new();
+	public List<GameSearchModel> GameResults { get; set; } = new();
+
+	[FromQuery]
+	[Display(Name = "Search Wiki")]
+	public bool PageSearch { get; set; } = false;
+
+	[FromQuery]
+	[Display(Name = "Search Forum Topics")]
+	public bool TopicSearch { get; set; } = false;
+
+	[FromQuery]
+	[Display(Name = "Search Forum Posts")]
+	public bool PostSearch { get; set; } = false;
+
+	[FromQuery]
+	[Display(Name = "Search Games")]
+	public bool GameSearch { get; set; } = true;
+
+	public int DisplayPageSize { get; set; } = PageSize;
+	public bool EnablePrev { get; set; } = false;
+	public bool EnableNext { get; set; } = false;
+
+	public async Task<IActionResult> OnGet()
+	{
+		if (!_db.Database.IsNpgsql())
+		{
+			ModelState.AddModelError("", "This feature is not currently available.");
+			return BadRequest(ModelState);
+		}
+
+		if (!ModelState.IsValid)
+		{
+			return Page();
+		}
+
+		// there is no dedicated method to check whether a regex string is valid, so we use try catch
+		try
+		{
+			Regex.Match("", SearchTerms);
+		}
+		catch (ArgumentException)
+		{
+			ModelState.AddModelError("SearchTerms", "Invalid Regular Expression.");
+			return Page();
+		}
+
+		if (!string.IsNullOrWhiteSpace(SearchTerms))
+		{
+			var seeRestricted = User.Has(PermissionTo.SeeRestrictedForums);
+			DisplayPageSize = PageSize;
+			if (new[] { PageSearch, TopicSearch, PostSearch, GameSearch }.Count(b => b) == 1)
+			{
+				DisplayPageSize = PageSizeSingle;
+			}
+
+			var skip = DisplayPageSize * (PageNumber - 1);
+			_db.Database.SetCommandTimeout(TimeSpan.FromSeconds(30));
+
+			if (PageSearch)
+			{
+				PageResults = await _db.WikiPages
+					.ThatAreNotDeleted()
+					.WithNoChildren()
+					.Where(w => Regex.IsMatch(w.PageName, SearchTerms))
+					.OrderBy(w => w.PageName)
+					.Skip(skip)
+					.Take(DisplayPageSize + 1)
+					.Select(w => new PageSearchModel(w.Markup.Substring(0, Math.Min(60, w.Markup.Length)), w.PageName))
+					.ToListAsync();
+			}
+
+			if (TopicSearch)
+			{
+				TopicResults = await _db.ForumTopics
+				.ExcludeRestricted(seeRestricted)
+				.Where(t => t.ForumId != SiteGlobalConstants.WorkbenchForumId && t.ForumId != SiteGlobalConstants.PlaygroundForumId && t.ForumId != SiteGlobalConstants.PublishedMoviesForumId && t.ForumId != SiteGlobalConstants.GrueFoodForumId)
+				.Where(t => Regex.IsMatch(t.Title, "(^|[^A-Za-z])" + SearchTerms))
+				.OrderByDescending(t => t.CreateTimestamp)
+				.Skip(skip)
+				.Take(DisplayPageSize + 1)
+				.Select(t => new TopicSearchModel(
+					t.Title,
+					t.Id,
+					t.Forum!.Name))
+				.ToListAsync();
+			}
+
+			if (PostSearch)
+			{
+				PostResults = await _db.ForumPosts
+				.ExcludeRestricted(seeRestricted)
+				.Where(p => Regex.IsMatch(p.Text, "(^|[^A-Za-z])" + SearchTerms))
```

Comment:
Does this generate sql?? If not, this brings back a half a million posts into memory and processes from there.  We MUST know this becomes a sql query, or else we can't do this

Reply:
From my tests it does. Here is an example from the log:
```
[20:05:58 INF] Executed DbCommand (6ms) [Parameters=[@__p_0='?', @__p_2='?' (DbType = Int32), @__p_1='?' (DbType = Int32)], CommandType='Text', CommandTimeout='30']
SELECT t.id, g0.code, t.display_name
FROM (
    SELECT g.id, g.display_name, g.system_id, (
        SELECT COUNT(*)::INT
        FROM publications AS p
        WHERE g.id = p.game_id) AS c, (
        SELECT COUNT(*)::INT
        FROM submissions AS s
        WHERE g.id = s.game_id) AS c0
    FROM games AS g
    WHERE g.display_name ~ ('(?p)' || @__p_0)
    ORDER BY (
        SELECT COUNT(*)::INT
        FROM publications AS p
        WHERE g.id = p.game_id) DESC, (
        SELECT COUNT(*)::INT
        FROM submissions AS s
        WHERE g.id = s.game_id) DESC, g.display_name
    LIMIT @__p_2 OFFSET @__p_1
) AS t
INNER JOIN game_systems AS g0 ON t.system_id = g0.id
ORDER BY t.c DESC, t.c0 DESC, t.display_name
```

The important part is the `WHERE g.display_name ~ ('(?p)' || @__p_0)` since the `~` is the regex operator in postgres.

Reply:
WOW. I'm impressed with EF.